pynsot/commands/cmd_protocols.py

```diff
@@ -0,0 +1,464 @@
+    ),
+    callback=callbacks.process_site_id,
+)
+@click.option(
+    '-t',
+    '--type',
+    metavar='TYPE',
+    type=types.NATURAL_KEY,
+    help='The type of the protocol.',
+    required=True,
+)
+@click.pass_context
+def add(ctx, auth_string, attributes, circuit, device, description, interface, site_id, type):
+    """
+    Add a new Protocol.
+
+    You must provide a Device hostname or ID using the -D/--device option.
+
+    You must provide the type of the protocol (e.g. OSPF, BGP, etc.)
+
+    If you wish to add attributes, you may specify the -a/--attributes
+    option once for each key/value pair.
+
+    """
+    data = ctx.params
+
+    if interface is None and circuit is None:
+        raise click.UsageError('Must have interface "-i" / "--interface" or circuit "-c" / "--circuit" populated')
+
+    if description is None:
+        data.pop('description')
+
+    ctx.obj.add(data)
+
+
+# List
+@cli.group(invoke_without_command=True)
+@click.option(
+    '-a',
+    '--attributes',
+    metavar='ATTRS',
+    help='A key/value pair attached to this Protocol (format: key=value).',
+    multiple=True,
+)
+@click.option(
+    '-u',
+    '--auth_string',
+    metavar='AUTH_STRING',
+    default='',
+    help='The authentication string (such as MD5 sum).',
+)
+@click.option(
+    '-c',
+    '--circuit',
+    metavar='CIRCUIT',
+    help='The circuit that this protocol is running over.',
+)
+@click.option(
+    '-d',
+    '--delimited',
+    is_flag=True,
+    help='Display set query results separated by commas vs. newlines.',
+    default=False,
+    show_default=True,
+)
+@click.option(
+    '-e',
+    '--description',
+    metavar='DESCRIPTION',
+    type=str,
+    help='Filter by Protocols matching this description.',
+)
+@click.option(
+    '-D',
+    '--device',
+    metavar='DEVICE',
+    type=types.NATURAL_KEY,
+    help=(
+        'Unique ID of the Device to which this Protocol is '
+        'running on.'
+    ),
+)
+@click.option(
+    '-g',
+    '--grep',
+    is_flag=True,
+    help='Display list results in a grep-friendly format.',
+    default=False,
+    show_default=True,
+)
+@click.option(
+    '-i',
+    '--id',
+    metavar='ID',
+    help='Unique ID of the Protocol being retrieved.',
+)
+@click.option(
+    '-I',
+    '--interface',
+    metavar='INTERFACE',
+    type=types.NATURAL_KEY,
+    help=(
+        'The Interface this Protocol is running on. Either interface'
+        'or circuit must be populated.'
+    ),
+)
+@click.option(
+    '-q',
+    '--query',
+    metavar='QUERY',
+    help='Perform a set query using Attributes and output matching Protocols.'
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    help='Unique ID of the Site this Protocol is under.',
+    callback=callbacks.process_site_id,
+)
+@click.option(
+    '-t',
+    '--type',
+    metavar='TYPE',
+    type=types.NATURAL_KEY,
+    help='The protocol type',
+    required=True,
```

Comment:
Type should not be required on `list`. By default we want to be able to do something like `nsot protocols list` to get all objects without any filters.

Reply:
The side effect here too is that it becomes impossible to even view a single object without also providing the type. e.g. `nsot protocols list -i 1` is not possible.